src/matchers.rs

```diff
@@ -58,6 +58,40 @@ impl Matcher for Child {
     }
 }
 
+/// Selects an array item by index.
+///
+/// If the index is negative, it references element len-abs(index).
+pub struct ArrayIndex {
+    index: i64,
+}
+
+impl ArrayIndex {
+    pub fn new(index: i64) -> Self {
+        ArrayIndex { index }
+    }
+}
+
+impl Matcher for ArrayIndex {
+    fn select<'a>(&self, node: &'a Value) -> Iter<'a> {
+        let len = if let Value::Array(a) = node {
+            a.len()
+        } else {
+            0
+        };
+        let idx = if self.index >= 0 {
+            self.index as usize
+        } else {
+            let abs = (-self.index) as usize;
+            if abs <= len {
```

Comment:
Is this approach to avoid overflow? If so, it would be good to include a unit test of that case.

(An alternative is to set `idx` to `len + self.index` if this expression is non-negative, but maybe that overflows for `usize` large negative `self.index`.)

Reply:
yeah I'm not happy with this code either; I did that just because self.index and len have incompatible types and rust detects overflow. 

but since json cannot have sparse arrays, I think it's perfectly fine to restrict the case of arrays with 2^63 elements :-)

PTAL

Reply:
The revised code looks good, thanks.